### PR TITLE
support controller to use custom workqueue

### PR DESCRIPTION
**Pull request by shaofan-hs**
…t custom Priority Queues

#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [x] N
- [ ] Y 

#### 2. What is the scope of this PR (e.g. component or file name):

Use NewRateLimitingQueueWithCustomQueue and MakeQueue to support custom priority queues.

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: controller/workqueue/priority_queue.go [hunk 1]
```diff
@@ -0,0 +1,315 @@
...
+func (q *PriorityQueue) Get() (item interface{}, shutdown bool) {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+
+	length := q.lenNoLock()
+	for length == 0 && !q.shuttingDown {
+		q.cond.Wait()
+	}
+	if length == 0 {
+		// We must be shutting down.
```
[the remark this answers is not in the document]
Comment (shaofan-hs) [reply controller/workqueue/priority_queue.go:233]: fixed

File: controller/workqueue/priority_queue.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+func (q *PriorityQueue) Len() int {
+	q.cond.L.Lock()
+	defer q.cond.L.Unlock()
+
+	return q.lenNoLock()
+}
+
+func (q *PriorityQueue) lenNoLock() int {
+	count := 0
+	for _, queue := range q.priorityQueue {
```
[the remark this answers is not in the document]
Comment (shaofan-hs) [reply controller/workqueue/priority_queue.go]: Yes, optimized.

File: controller/workqueue/priority_queue.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+	defer q.cond.L.Unlock()
+	if q.shuttingDown {
+		return
+	}
+	if q.dirty.has(item) {
+		return
+	}
+
+	priority := q.getPriorityFunc(item)
+	if priority < 0 || priority > q.maxPriority {
```
[the remark this answers is not in the document]
Comment (shaofan-hs) [reply controller/workqueue/priority_queue.go]: fixed

File: controller/workqueue/defaults.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	return func(item interface{}) int {
+		clientObject, ok := item.(client.Object)
+		if !ok {
+			return defaultWorkQueuePriority
+		}
+
+		var (
+			checkNamespace     = false
```
[the remark this answers is not in the document]
Comment (shaofan-hs) [reply controller/workqueue/defaults.go]: fixed

**Review by wu8685 — APPROVED** [review]
lgtm

**Pull request merged**